### PR TITLE
Add Parallelization when handling Domain Events with Pull Delivery

### PR DESCRIPTION
**Pull request by Le-Merle**
## Description of changes
When pulling events, the call to EventGrid asked for a single event, then processed the event, then called back to report (acknowledge, reject, release) before starting over and asking for the next single event.

This PR changes the behavior by accepting a new configuration on the subscription to allow running multiple domain event handlers in parallel. The call to EventGrid will request as many events as possible to fill the available space, improving the efficiency per roundtrip. A similar logic applies to the callbacks, where multiple token locks can be sent back at once.

For comparison, here are the results of running PullDeliveryTests.TestHeavyLoadOfEvents on the old implementation versus the new when configured for 500 handlers at once.

PREVIOUS IMPLEMENTATION
Processed 10 out of 10 events in 2664.5465 ms (3.75 event/second) with handler execution time at 250 ms
Processed 100 out of 100 events in 26631.8249 ms (3.75 event/second) with handler execution time at 250 ms
Processed 1000 out of 1000 events in 266582.5875 ms (3.75 event/second) with handler execution time at 250 ms
(Processing 10000 events reached the 5 minutes limit)

NEW IMPLEMENTATION
Processed 10 out of 10 events in 267.6413 ms (37.36 event/second) with handler execution time at 250 ms
Processed 100 out of 100 events in 358.0868 ms (279.26 event/second) with handler execution time at 250 ms
Processed 1000 out of 1000 events in 1087.8638 ms (919.23 event/second) with handler execution time at 250 ms
Processed 10000 out of 10000 events in 7067.7937 ms (1414.87 event/second) with handler execution time at 250 ms

In short, handling 100 events with the old system took 26s, and with the new system it takes 7s to handle 10k events.

## Breaking changes
There is a new configuration option on the subscription to control the degree of parallelism, but the default value is set to 1, so it should be retro compatible.

## Additional checks
- [ ] Updated the documentation of the project to reflect the changes
- [x] Added new tests that cover the code changes

### CODE REVIEW

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPropagationSubscriptionOptions.cs [hunk 1]
```diff
@@ -15,4 +15,6 @@ public class EventPropagationSubscriptionOptions
     public string TopicName { get; set; } = string.Empty;
 
     public string SubscriptionName { get; set; } = string.Empty;
+
+    public int MaxDop { get; set; } = 1;
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPropagationSubscriptionOptions.cs:19]:
```suggestion
    public int MaxDegreeOfParallelism { get; set; } = 1;
```

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 2]
```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
...
+                this.FeedEventsIntoChannel(cancellationToken),
+                this.HandleCompletedEventHandlers(cancellationToken),
+                this.AcknowledgeEvents(cancellationToken),
+                this.ReleaseEvents(cancellationToken),
+                this.RejectEvents(cancellationToken));
+        }
+
+        private async Task FeedEventsIntoChannel(CancellationToken cancellationToken)
+        {
+            using var scope = this._serviceScopeFactory.CreateScope();
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:75]:
```suggestion
            await using var scope = this._serviceScopeFactory.CreateAsyncScope();
```

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 3]
```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
...
-    }
+        private static async Task<List<string>> ReadCurrentChannelContent(Channel<string> channel, CancellationToken cancellationToken)
+        {
+            await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false);
 
-    private static async Task RejectEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.RejectCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
+            var currentChannelCount = channel.Reader.Count;
+            var results = new List<string>();
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:169]: _I don't know if this is important for perf (would be better to test it)._

I think you could preallocate the list in the calling method with the expected capacity. Before calling `ReadCurrentChannelContent`, you can clear it. This way, you only allocates 3 lists for the whole lifetime of the app, instead of one list each time you call this method.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 4]
```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
...
-                foreach (var (cloudEvent, lockToken) in bundles)
+                try
+                {
+                    await this._handlerTasksChannel.Writer.WaitToWriteAsync(cancellationToken).ConfigureAwait(false);
+
+                    var bundles = await this._eventGridTopicSubscription.Client.ReceiveCloudEventsAsync(this._eventGridTopicSubscription.TopicName, this._eventGridTopicSubscription.SubscriptionName, this._eventGridTopicSubscription.MaxHandlerDop - this._handlerTasksChannel.Reader.Count, cancellationToken).ConfigureAwait(false);
+
+                    foreach (var bundle in bundles)
+                    {
+                        await this._handlerTasksChannel.Writer.WriteAsync(HandleEventAsync(cloudEventHandler, bundle.Event, bundle.LockToken, cancellationToken), cancellationToken).ConfigureAwait(false);
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:88]: If `HandleEventAsync` is blocking, you are not writing other items.

I wonder if a better implementation could be
- Write the `bundle` to the Channel directly
- Create another method with the following code

```
await Parallel.ForEachAsync(channel.Reader.ReadAllAsync(), (item, cancellationToken) =>
{
    await cloudEventHandler.HandleCloudEventAsync(cloudEvent, cancellationToken).ConfigureAwait(false);
    // Write result to acknoledge/rejected/release channel (similar to the method HandleCompletedEventHandlers)
});
```

Also, I think you can remove `TaskBoundedChannel<T>`.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests/PullDeliveryTests.cs [hunk 5]
```diff
@@ -47,12 +49,55 @@ public async Task TestPublishAndReceiveEvent()
         await runTask;
     }
 
+    [Theory]
+    [InlineData(10)]
+    [InlineData(100)]
+    [InlineData(1000)]
+    [InlineData(10000)]
+    public async Task TestHeavyLoadOfEvents(int eventCount)
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests/PullDeliveryTests.cs:57]: I don't think the emulator can return multiple events at the moment. I don't know if this impacts the tests.

Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests/PullDeliveryTests.cs:57]: It probably lowers the overall performance gains, but the test only makes sure that we are running faster than serially (aka at a faster event/second rate than 1000ms / "slow task" delay).

In the end I don't think it could become flaky unless we used a lower "slow task" delay and reading from the emulator becomes the bottleneck. With ~250ms it should be safe.

Still it could be a nice improvement on the emulator to support this!

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 4 above)
Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:88]: I was hesitant to go the `Parallel.ForEachAsync` route instead of simply running Tasks, for 2 main reasons:

1. I am not sure which is more appropriate resource/performance wise, or if they are simply equivalent
2. I do not want to prefetch events if there is no room to start working on them right away in order to scale better when multiple pods are consuming the same subscription, and `Parallel.ForEachAsync` does not offer a way to know how many tasks are currently running

To work around #2, I would have to keep a localized count of running tasks that gets incremented/decremented in a thread safe way, and I would need a way to notify the EventGrid reader that there is now more room available so it can go fetch more events. A channel fits those requirements, so I went ahead and wrapped my solution inside one to simplify the logic when using it.

I am very open to discussing potential solutions for this, I went with something that works, but I'm not entirely happy with the partial channel as it stands.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 3 above)
Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:169]: The way it currently works, it will grab whatever currently resides within the channel and send it all at once to the acknowledge/release/reject endpoint. There is no way to know how many events are currently sitting in there.

However, it might be more prudent to limit the size of a single batch. There might be a limit on the amount of lockTokens being sent at once, and overall it would allow for reusable lists with fixed sizes, since we know we can't outgrow it. I'll implement this.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 4 above)
Comment (meziantou) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:88]: > I am not sure which is more appropriate resource/performance wise, or if they are simply equivalent

You will need to measure :)

For the second point,
- Number of events in the channel is already available
- For the number of events being processed, you can add a field and use `Interlocked.Increment` in `Parallel.ForEachAsync` and `Decrement` at the end of the loop or after acknowledge/reject/release (just be sure this is always called)
- You may also need a signal to indicate when an item has completed, so maybe a manualResetEvent or something similar, in addition to `Writer.WaitToWriteAsync()`.

Comment (meziantou) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:88]: BTW, maybe a simpler implementation would not require a Channel for the events. Here's a small POC, not battle tested.

````
// Simulate fetching events from Azure Event Grid
int lastEvents = 0;
int[] FetchFromEventGrid(int numbers)
{
    var result = Enumerable.Range(lastEvents, numbers).ToArray();
    lastEvents += numbers;
    return result;
}

// Local loop to fetch events from Azure
int currentlyProcessing = 0;
ManualResetEventSlim manualResetEventSlim = new ManualResetEventSlim(initialState: true);
async IAsyncEnumerable<int> FetchLoop()
{
    var maxFetch = 5;
    while (true)
    {
        manualResetEventSlim.Wait();
        var number = maxFetch - currentlyProcessing;
        if (number <= 0)
            continue; // TODO be smart

        var data = FetchFromEventGrid(number);
        Console.WriteLine($"Fetch {data.Length} items");
        foreach (var item in data)
        {
            yield return item;
        }
    }
}

// Processing loop
var parallelOptions = new ParallelOptions { MaxDegreeOfParallelism = 5 };
await Parallel.ForEachAsync(FetchLoop(), async (item, cancellationToken) =>
{
    Interlocked.Increment(ref currentlyProcessing);
    Console.WriteLine(item);
    await Task.Delay(1000);
    Interlocked.Decrement(ref currentlyProcessing);
    manualResetEventSlim.Set();
});
````

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 6]
```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
...
     {
-        using var scope = this._serviceScopeFactory.CreateScope();
-        var cloudEventHandler = scope.ServiceProvider.GetRequiredService<ICloudEventHandler>();
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+        private readonly ILogger<EventPullerService> _logger;
+        private readonly EventGridTopicSubscription _eventGridTopicSubscription;
 
-        while (!stoppingToken.IsCancellationRequested)
+        private readonly Channel<Task<EventHandlingResult>> _handlerTasksChannel;
+        private readonly Channel<string> _acknowledgeEventChannel = Channel.CreateUnbounded<string>();
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:46]: I think these channels should be bounded. Most of the time it should not be a problem, but if you process messages to quickly, or if there is an issue, it can create back-pressure. Also processed messages may timeout and will be reprocessed.
A bounded channel may be a safety-net

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 4 above)
Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:88]: I've spent quite a bit of time playing around different ways to implement this today.

The simplified version you suggest (with the minor modification of using a `TaskCompletionSource` instead of the thread blocking `ManualResetEventSlim`) works, but using `Parallel.ForEachAsync` consistently produces a slightly lower throughput than simply running the tasks.

However, the difference is negligible versus the task-only version, and I cannot be certain of what it will look like on a production-grade server versus my own machine. `Parallel.ForEachAsync` apparently instructs the OS to dispatch threads on different CPU cores as much as possible, versus the basic Tasks where they let the OS pick where to run them, which makes it rather difficult to compare when not on the final location with the real-world specs. From what I've read, `Parallel.ForEachAsync` is more suited to CPU-bound tasks where having a good spread across the CPU cores will yield benefits, while Tasks will work very well on network-bound tasks since waiting does not consume anything. In any case, this type of micro-optimization is highly dependant on the work being done in the handler, and we don't have control over that by definition, so there shouldn't be a single "correct" solution here performance-wise.

I'll go with the simplified `Parallel.ForEachAsync` version as it doesn't imply a homemade custom class.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 6 above)
Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs]: Switched to bounded channels with a limit of 5000.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs (hunk 3 above)
Comment (Le-Merle) [reply src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs]: I changed the way the reading occurs, it will now read up to the current count when starting to read but it returns an `IEnumerable` that is directly passed to the client.

Since the client receives an IEnumerable, I am assuming they will handle proper partitionning of a number of tokens that is too high. This also has the benefit of removing the allocation of intermediary lists.

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 7]
```diff
@@ -20,81 +20,199 @@ internal class EventPullerService : BackgroundService
...
+            {
+                this._taskCompletionSource.TrySetResult();
+                this._taskCompletionSource = new TaskCompletionSource();
+            }
+        }
 
-    private static async Task RejectEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.RejectCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
+        private Task WaitForAvailableHandler(CancellationToken cancellationToken)
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:187]: Maybe this can be a `ValueTask` to reduce allocation in the best case?

If this is a bottleneck, which I'm not really not sure, a potential optimization would be to `IValueTaskSource` in addition to the `ValueTask`

File: src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs [hunk 8]
```diff
@@ -20,81 +20,199 @@ internal class EventPullerService : BackgroundService
...
+            });
+        }
+
+        private async IAsyncEnumerable<EventGridClientAdapter.EventGridClientAdapter.EventBundle> StreamEventGridEvents([EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                await this.WaitForAvailableHandler(cancellationToken).ConfigureAwait(false);
+
+                var availableHandlers = this._eventGridTopicSubscription.MaxHandlerDop - this._handlersInProgress;
```
Comment (meziantou) [inline src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs:105]: I'm really not sure this is needed

```suggestion
                var availableHandlers = this._eventGridTopicSubscription.MaxHandlerDop - Volatile.Read(ref this._handlersInProgress);
```

**Review by meziantou — APPROVED** [review]

**Review by meziantou — APPROVED** [review]
